google/cloud/bigtable/mutations_batcher.py

```diff
@@ -46,61 +147,214 @@ class MutationsBatcher:
+          - flush_interval: Automatically flush every flush_interval seconds
+          - flush_limit_count: Flush immediately after flush_limit_count mutations are added.
+              If None, this limit is ignored.
+          - flush_limit_bytes: Flush immediately after flush_limit_bytes bytes are added.
+              If None, this limit is ignored.
+          - flow_control_max_count: Maximum number of inflight mutations.
+              If None, this limit is ignored.
+          - flow_control_max_bytes: Maximum number of inflight bytes.
+              If None, this limit is ignored.
+        """
+        atexit.register(self._on_exit)
+        self.closed: bool = False
+        self._table = table
+        self._staged_mutations: list[RowMutationEntry] = []
+        self._staged_count, self._staged_bytes = 0, 0
+        self._flow_control = _FlowControl(
+            flow_control_max_count, flow_control_max_bytes
+        )
+        self._flush_limit_bytes = (
+            flush_limit_bytes if flush_limit_bytes is not None else float("inf")
+        )
+        self._flush_limit_count = (
+            flush_limit_count if flush_limit_count is not None else float("inf")
+        )
+        self.exceptions: list[FailedMutationEntryError] = []
+        self._flush_timer_task: asyncio.Task[None] = asyncio.create_task(
+            self._flush_timer(flush_interval)
+        )
+        # create noop previous flush task to avoid None checks
+        self._prev_flush: asyncio.Task[None] = asyncio.create_task(asyncio.sleep(0))
+        # MutationExceptionGroup reports number of successful entries along with failures
+        self._entries_processed_since_last_raise: int = 0
+
+    async def _flush_timer(self, interval: float | None):
+        """
+        Triggers new flush tasks every `interval` seconds
+        """
+        if interval is None:
+            return
+        while not self.closed:
+            await asyncio.sleep(interval)
+            # add new flush task to list
+            if not self.closed and self._staged_mutations:
+                self._schedule_flush()
 
-    async def append(self, row_key: str | bytes, mutation: Mutation | list[Mutation]):
+    def append(self, mutations: RowMutationEntry):
+        """
+        Add a new set of mutations to the internal queue
         """
-        Add a new mutation to the internal queue
+        if self.closed:
+            raise RuntimeError("Cannot append to closed MutationsBatcher")
+        size = mutations.size()
+        if size > self._flow_control.max_mutation_bytes:
+            raise ValueError(
+                f"Mutation size {size} exceeds flow_control_max_bytes: {self._flow_control.max_mutation_bytes}"
+            )
+        if len(mutations.mutations) > self._flow_control.max_mutation_count:
+            raise ValueError(
+                f"Mutation count {len(mutations.mutations)} exceeds flow_control_max_count: {self._flow_control.max_mutation_count}"
```

Comment:
I think we  want flow control to be advisory rather then a hard limit. Its meant to limit the throughput of an individual client rather than prevent traffic. Having said that, I think there are separate limits that need to be enforced (100k mutations per entry and 100k entries.

So batch flush limits should be loose and force a flush even after an overflow. But the hard limits should be enforced at the model level

Reply:
Ok, I changed it so that the flow control will allow oversized entries as a single-entry request, and I added some extra checks for very large entries and requests

Is there a hard limit to the bytes size of an individual entry/request I should be checking for as well as well?